---
title: "Simulating cutthroat-rainbow trout hybrids and passing them to NewHybrids"
author: "Eric C. Anderson"
date: "Last Updated: `r Sys.Date()`"
output:
  html_document:
    df_print: paged
    toc: true
    toc_float: true
bibliography: "`r here::here('references.bib')`"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(message = FALSE)
knitr::opts_knit$set(root.dir = here::here())
start_time <- Sys.time()
```

```{r rmd_snakemake_hacks, include=FALSE}
# here is a little code to make it easy to work interactively with this
# Rmarkdown document to try to find problems.  WHen you run it in snakemake,
# it save the snakemake S4 class object.  When there is no snakemake object,
# it reads it from the file that was previously saved.
rsdir <- ".rmd_snakemake"
dir.create(rsdir, showWarnings = FALSE)
rdsfile <- file.path(rsdir, "admix-fracts.rds")
if(exists("snakemake")) {
  saveRDS(snakemake, file = rdsfile)
} else {
  snakemake <- readRDS(file = rdsfile)
}
```



Load the libraries.
```{r, warning=FALSE, message=FALSE}
library(tidyverse)
library(cowplot)
library(gscramble)
library(readxl)

```



@rizza2023limited used genetic markers to identify hybrid individuals between steelhead
trout and cutthroat trout.  They used the program NewHybrids [@anderson2002model], to
do so.

They had 65 diagnostic or nearly diagnostic markers at their disposal, however they
used just 30 of those to ensure that they didn't need to worry about physical linkage.

They hybrid categories they considered when doing inference with NewHybrids were:

| Category | P(CCT/CCT)   |  P(CCT/SH or SH/CCT) |  P(SH/SH) |
|:---------: | :------: | :--------:  | :---------: |
| CCT  |  1.0000  |  0.0000  |  0.0000 |
| SH  |  0.0000  |  0.0000  |  1.0000 |
| F1  |  0.0000  |  1.0000  |  0.0000 |
| F2  |  0.2500  |  0.5000  |  0.2500 |
| BxC  |  0.5000  |  0.5000  |  0.0000 |
| BxSH  |  0.0000  |  0.5000  |  0.5000 |
| BxC2  |  0.7500  |  0.2500  |  0.0000 |
| BxSH2  |  0.0000  |  0.2500  |  0.7500 |
| BxCxBxC2  |  0.5625  |  0.3750  |  0.0625 |
| BxSHxBxSH2  | 0.0625 |  0.3750  |  0.5625 |
            

We are curious about where a variety of different hybrid categories might
get assigned---both the ones specified above, as well as other ones,
such as the `F2 x Bx` or `F2 x Bx2` sort of category. In other words we
want to know about the accuracy that we can expect.

And we also want to investigate the differences betwween using 30 vs
all 65 markers---do we really overestimate certainty, etc?



# Making a Map

These markers were mapped by @rizza2023limited to the Omyk_v1.0 genome.  A linkage map
for rainbow trout / steelhead from 57K affymetrix chip markers is available.  We 
make a sex-averaged map by mapping these 57K markers to Omyk_v1.0 and averaging recombination
fractions over males and females and over the two different families.

Since we are talking about recombination between hybrids, the rates might be
different (likely lower) than in these particular families of rainbow trout, but
this is a good start.  

```{r}
# read in the data files we have
map <- read_excel("data/cutts/rainbow_trout_genetic_map_AG_TL.xlsx")
affy_meta <- read_excel("data/cutts/men12337-sup-0002-appendixs2.xlsx")

# get the flanking region for each Affy SNP ID and replace the [X/Y] with X
# and also count the position from the beginning of the sequence that [X/Y] construct
# falls in the sequence.
affy_seqs <- affy_meta %>%
  select(`Affy SNP ID`, Flank) %>%
  mutate(
    start = str_locate(Flank, pattern = "\\[./.\\]")[,1],
    end = str_locate(Flank, pattern = "\\[./.\\]")[,2],
    repl = str_sub(Flank, start = start+1, end = start+1),
    newflank = str_replace(Flank, pattern = "\\[./.\\]", replacement = repl),
    oldlength = str_length(Flank),
    newlength = str_length(newflank)
  )

# now, write those to a fasta
dir.create("intermediates", showWarnings = FALSE)
tmp <- paste(">", affy_seqs$`Affy SNP ID`, "\n", affy_seqs$newflank, sep = "")
cat(tmp, sep = "\n", file = "intermediates/affy_flanks.fasta")
```

Now, we are going to map that to the Omyk_v1.0.  To make this work, other users
will have to download that and then they will have to have bwa on their system
and then bwa index it and update the GPATH variable in the
shell block below.
```sh
GPATH=/Users/eriq/Documents/UnsyncedData/omyV6-genome/omyV6Chr.fasta
conda activate bioinf
bwa mem $GPATH intermediates/affy_flanks.fasta > intermediates/affy_flanks.sam

# sort it and then make a file to read into R
samtools view -u intermediates/affy_flanks.sam | samtools sort  | samtools view | awk 'BEGIN {OFS="\t"; print "id","samflag", "chrom", "map_pos", "qual"} /^Affx/ && $5==60 {print $1, $2, $3, $4, $5}' > intermediates/affy_flanks.tsv
```

Now, read that back into R so we can find the positions of those.
```{r}
affy_sam <- read_tsv("intermediates/affy_flanks.tsv")

affy_pos <- affy_sam %>%
  left_join(affy_seqs, by = join_by(id == `Affy SNP ID`)) %>%
  mutate(
    pos = case_when(
      samflag == 0 ~ map_pos + start - 1L,
      samflag == 16 ~ map_pos + newlength - start, # might be off by 1 here, but for our purposes no matter
      TRUE ~ NA_integer_
    )
  ) %>%
  filter(!is.na(pos)) %>%
  select(id, chrom, pos)
```

Now, clean up the map a bit.  First off, just fill the NAs with the previous
row's values:
```{r}
ave_map <- map %>%
  fill(-c(Marker, omy)) %>%
  mutate(averaged = (`AG-male` + `AG-female` + `TL-male` + `TL-female`) / 4)

```

Then, join that to the physical positions:
```{r}
phys_map <- affy_pos %>%
  inner_join(ave_map %>% select(Marker, averaged), by = join_by(id == Marker))
```

Now have a look at that:
```{r}
ggplot(phys_map, aes(x = pos, y = averaged)) +
  geom_line() + 
  facet_wrap(~ chrom, ncol = 5)
```

So, that is kind of weird looking.  There is likely some misassembly, with
some things that belong in the middle of the chromosome ending up near the front
of it in the physical assembly, possibly.


Also, let's focus on omy05:
```{r}
phys_map %>%
  filter(chrom == "omy05") %>%
  ggplot(aes(x = pos, y = averaged)) +
  geom_line() + 
  facet_wrap(~ chrom, ncol = 5)
```

So that might be because one family has a parent polymorphic for the inversion.  

Anyhoo, we are going to need to clean this stuff up.  
```{r}
oldie <- phys_map
for(i in 1:100) {
newie <- oldie %>%
  filter(lead(averaged) > averaged)
oldie <- newie  
}

phys_map_filt <- newie

ggplot(phys_map_filt, aes(x = pos, y = averaged)) +
  geom_line(colour = "blue") + 
  facet_wrap(~ chrom, ncol = 5)
```

OK! That looks suitably clean.  

Now, I am going to drop any markers at which a change from the previous has not
occurred, and then we can make a plink map out of it.
```{r}
plink_map_tib <- phys_map_filt %>%
  mutate(diff = averaged - lag(averaged)) %>%
  filter(!is.na(diff) & diff > 0.0) %>%
  mutate(morgans = averaged / 100) %>%
  select(chrom, id, morgans, pos)

mapfile <- "intermediates/affy-plink-morgans-map.gz"
write_tsv(plink_map_tib, col_names = FALSE, file = mapfile)

cutt_RecRates <- plink_map2rec_rates(mapfile, use_morgans = TRUE)
```

# Making the markers object

Read in the markers and genotypes:
```{r}
# just put each variant in the center of where it mapped. We don't
# need to be right on it for this, and 20 or 30 bp here and there
# will make no difference.
cutts_marker_meta <- read_csv("data/cutts/cutt-markers-mapped.csv") %>%
  rename(
    chrom = chromosome,
    variant_id = marker_name
  ) %>%
  mutate(pos = floor((start + end) / 2)) %>%
  select(chrom, pos, variant_id) %>%
  arrange(chrom, pos)
  
  
# the genotypes are basically coded by species.  So let's clean it up a bit and filter
# it down to the 65 loci in cutts_marker_meta
cutt_geno_long <- read_csv("data/cutts/Loci_96_Indiv_876.csv") %>%
  pivot_longer(cols = -c(no, ID, `Hybrid Class`), names_to = "locus", values_to = "geno") %>%
  semi_join(cutts_marker_meta, by = join_by(locus == variant_id)) %>%
  mutate(
    `1` = case_match(
      geno,
      "M" ~ "M",
      "C" ~ "C",
      "H" ~ "C",
      "G" ~ NA_character_
    ),
    `2` = case_match(
      geno,
      "M" ~ "M",
      "C" ~ "C",
      "H" ~ "M",
      "G" ~ NA_character_
    )
  ) %>%
  select(-geno) %>%
  pivot_longer(cols = c(`1`, `2`), names_to = "gene_copy", values_to = "allele")

# then pivot that wider and make sure things are in the correct
# order, keeping only the ones that were deemed pure SH and CCT
cutt_geno_wide <- cutt_geno_long %>%
  pivot_wider(
    names_from = c(locus, gene_copy),
    names_sep = "_",
    values_from = allele
  ) %>%
  rename(
    group = `Hybrid Class`,
    indiv = ID
  ) %>%
  filter(group %in% c("SH", "CCT"))

# this we will use in gscramble
cutt_I_meta <- cutt_geno_wide %>%
  select(group, indiv) %>%
  mutate(indiv = as.character(indiv))

# rename the marker meta to something more familiar (like the data
# from the gscramble package)
cutt_M_meta <- cutts_marker_meta

# now, also get the genotypes in the format needed
cgw2 <- cutt_geno_wide[, paste(rep(cutt_M_meta$variant_id, each = 2), c(1, 2), sep ="_")] %>%
  as.matrix()
colnames(cgw2) <- NULL
rownames(cgw2) <- cutt_I_meta$indiv

cutt_Geno <- cgw2

```

So, that is our 65 locus data.  The objects we will move forward with from those
are:

```{r, eval=FALSE}
cutt_I_meta
cutt_M_meta
cutt_Geno
cutt_RecRates
```

Let's make sure that all the markers are within the map. With this and some other
sleuthing, we see that one marker is just outside of the map.  So, will will adjust
the map.
```{r}
cutt_RecRates %>%
  left_join(cutt_M_meta, by = join_by(chrom), relationship = "many-to-many") %>%
  group_by(chrom) %>%
  filter(pos > max(end_pos)) %>%
  tail()
```

```{r}
cutt_RecRates <- cutt_RecRates %>%
  mutate(
    chrom_len = ifelse(chrom == "omy14", 70770590, chrom_len),
    end_pos = ifelse(chrom == "omy14" & end_pos == 68317429, 70770591, end_pos)
  )
```

Now, the map and markers should work well together.

# Making a GSP

Let's make a GSP that will let us get 1 or two of each of those categories so that
we end up with a sample that looks a little like what was inferred, in terms of
numbers of a fish of different categories.  

In fact, looking at Table 1 in @rizza2023limited, in order to simulate data that
look like what we observed, we would have something like:

- 15 F1
- 1 F2
- 1 of the weird FNxBxC fish
- 6 BxC
- 6 BxC2

So, we will shoot for something that is approximately like that. And let's add some
BxC3's in there too, for good measure.  We also want to have enough individuals that
we get a good sample of each of the categories. So, something approximately like:

- 15 F1
- 4 F2
- 3 BxC x BxC2
- 8 BxC
- 8 BxC2
- 5 BxC3

We need to put together some pedigrees that will get us this.  We do that by
modifying some of the ones made by `create_gsp()`.  

So, 1 pedigree with 8 BxC2's and 1 pedigree with 4 BxCs can be joined up. Mating
2 BxC2's with 2 BxC's to get 4 BxC3,
leaving us with 6 BxC2's and 2 BxCs.


Let's make that pediguree.
```{r}
BxC2_8 <- create_GSP(pop1 = "c", pop2 = "s", F1B2 = TRUE)
BxC_4 <- create_GSP(pop1 = "c", pop2 = "s", F1B = TRUE) %>%
  renumber_GSP(add = 15)

# now, we have 23 the offspring of 12 and 21, so i is a BxC x BxC2's.
# then we also mate 15 to 26---the offspring of 2 new cutt founders, 24 and 25---to 
# get 8 BxC3's with id 27.
add_em <- tibble(
  ind = c(23, 24, 25, 26, 27),
  par1 = c(12, NA, NA, 24, 15),
  par2 = c(21, NA, NA, 25, 26),
  ipar1 = c(4, NA, NA, 2, 4),
  ipar2 = c(4, NA, NA, 2, 4),
  sample = c("s23", NA, NA, NA, "s27"),
  osample = c(4L, NA, NA, NA, 4L),
  hpop1 = c(NA, "c", "c", NA, NA),
  hpop2 = c(NA, "c", "c", NA, NA),
  hap1 = c(NA, "24a", "25a", NA, NA),
  hap2 = c(NA, "24b", "25b", NA, NA)
)

newGSP <- bind_rows(BxC2_8, BxC_4) %>% 
  mutate(
    sample = case_when(
      ind %in% c(12, 15, 21) ~ NA,
      TRUE ~ sample
      ),
    osample = case_when(
      ind %in% c(12, 15, 21) ~ NA,
      TRUE ~ osample
      ),
  ) %>%
  bind_rows(add_em)

# let's quickly plot that to make sure it is looking correct
dotpaths <- gsp2dot(newGSP)

knitr::include_graphics(dotpaths["svg"])
```

What we get out of that are:
```{r}
newGSP_key <- tibble(
  name = "newGSP",
  samples = c("s13", "s14", "s27", "s22", "s23"),
  n = c(2, 2, 4, 2, 4),
  type = c("BxC2", "BxC2", "BxC3", "BxC", "BxCxBxC2")
)

newGSP_key
```

Great.  That covers most of the difficult cases.  What we need now are
about 16 F1s and a few F2's and some BxC's.  We can just do however
many replicates of each `create_GSP()` that gives us those.  

```{r}
# 4 backcrosses
BxC_4 <- create_GSP("c", "s", F1B = TRUE)

# 2 F1's
F1_2 <- create_GSP("c", "s", F1 = TRUE)

# 4 F2's
F2_4 <- create_GSP("c", "s", F2 = TRUE)
```

And we just need to make reppop's for each of those.

```{r}
# this is for the GSPs we do just a single rep of (F2, newGSP, BxC_4)
single_rep <- tibble(
  index = c(1L, 1L),  # note, index has be an integer.  Hence the "L" on each number
  pop = c("c", "s"),
  group = c("CCT", "SH")
)

# we will get 16 F1's by repping the F1_2 8 times
eight_rep <- tibble(index = rep(1:8, each = 2)) %>%
  mutate(
    pop = rep(c("c", "s"), length.out = n()),
    group = rep(c("CCT", "SH"), length.out = n())
  )
```

With these we can make our input tibble:
```{r}
InputTibble <- tibble(
  gpp = list(newGSP, BxC_4, F2_4, F1_2),
  reppop = list(single_rep, single_rep, single_rep, eight_rep)
)
```

# Looking at some simulated segments

Let's simulate some segments through these genomes and have a look:
```{r}
set.seed(122) # for reproducibility
Segments <- segregate(
  request = InputTibble,
  RR = cutt_RecRates
)
```

Note that this has created 38 simulated hybrid individuals:
```{r}
Segments %>% 
  count(gpp, index, ped_sample_id, samp_index)
```

Let's see if we can plot those:
```{r}
g <- plot_simulated_chromomsome_segments(Segments, cutt_RecRates)

#ggsave(g, filename = "/tmp/segs.pdf", width = 30, height = 30)
```
That only looks good as a high-res PDF.  But it shows that this is doing what
we think it should be doing.


# A quick trial of simulating some genotypes

```{r}
Segments2 <- segregate(
  request = InputTibble,
  RR = cutt_RecRates,
  MM = cutt_M_meta
)

Markers <- segments2markers(
  Segs = Segments2,
  Im = cutt_I_meta,
  Mm = cutt_M_meta,
  G = cutt_Geno
)
```

We are close to ready to do a big set of simulations with these. 

## Literature Cited


